### PR TITLE
Rename mosaic layout

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

This gives us the option to in the future combine *tight_layout* and *constrained_layout* into a single *layout* kwarg.

Even though this is provisional so I do not think we need to do a deprecation cycle, we may want to hold adding the layout kwarg until 3.6?

I took this opportunity to update the tutorial to use `fig.subfigures()` as well!

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: doc/api/next_api_changes/behavior/99999-tac.rst [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Rename fist arg to subplot_mosaic
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Both `.FigureBase.subplot_mosaic`, and `.pyplot.subplot_mosaic` have had the
+first position argument renamed from *layout* to *mosaic*.  This is because we
+are considering to consolidate *constrained_layout* and *tight_layout* keyword
+arguments in the Figure creation functions of `pyplot` into a single *layout*
```
Comment (jklymak) [inline doc/api/next_api_changes/behavior/99999-tac.rst:7]:
```suggestion
arguments in the Figure creation functions of ``pyplot`` into a single *layout*
```

Also rename this file `20150.rst`

Comment (tacaswell) [reply doc/api/next_api_changes/behavior/99999-tac.rst]: Went with fixing the xref rather than making it a code literal.

**Review by jklymak — APPROVED** [review]
Looks good to me.  Even though we may not decide to use "layout" for the other kwarg, changing this name to something less generic seems helpful.

**Review by timhoffm — APPROVED** [review]
I'm not 100% happy with the name `mosaic`. In particular when this is picked up as a variable name
```
mosaic = """AA
            BC"""
```
it feels a little vague. But I don't have a bette name.

Side remark: The short-hand "AA;BC" should be mentioned in the tutorial as well.

**Comment by timhoffm** [comment]
I'm unclear if we should introduce or pre-announce this in 3.4.2. The faster we make the change the better.

Similarly, we may still consider pushing for `layout` in 3.5. The sooner we change away from `constrained_layout=True` the less people have to re-learn. I assume `constrained_layout` is still in the phase of picking up. That faster change would come at the cost of the (likely) one person who uses `subplots_mosaic` by kwarg only. I'd take the risk of that very unlikely break to move the majority to a better `layout` parameter API earlier.

**Comment by tacaswell** [comment]
I am :+1: on backporting this for 3.4.2, contrary to my normally conservative position on these sorts of things.  My reasoning is that given that we labeled this as provisional, then we should have the courage of our convictions and treat it as provisional!  Setting that precedent is probably a good thing and this seems like a low-stakes case to try that in.

I did consider leaving the variable as mosaic as the variable the tutorial, but went with changing it everywhere.  Would going with variables like

```
AABC = "AA;BC"
```

be better?

I agree we should show of the short hand, I'll add that section once we agree on the variable names.

----

I would also propose that with this name change we make mosaic non-provisional for 3.5, but I'll do that after this is in.

**Comment by timhoffm** [comment]
> Would going with variables like
> 
> ```
> AABC = "AA;BC"
> ```
>
> be better?

Um, no. That's rather cryptic and it's not a pattern that scales. FWIW the single-line pattern should be inlined most of the time: `subplot_mosaic('AA;BC')`.

What about alternative names:
- arrangement
- formation
- structure
- placement
- zones

**Comment by WeatherGod** [comment]
My 2 cents, I thought "mosaic" was fine. IIRC, the context of that name was
to have something similar to "patchwork" so people might think it is
related.

**Comment by timhoffm** [comment]
@WeatherGod not sure you got the context right. The function name will stay. It's only the name of the first argument.

Up to now:
`subplot_mosaic(layout, ...)`.

Future:
`subplot_mosaic(mosaic, ...)` or 
`subplot_mosaic(arrangement, ...)` or ...

**Comment by anntzer** [comment]
fwiw I think using `mosaic` as param name is just fine.

**Comment by jklymak** [comment]
Not sure if this should be backported to 3.4.2?

**Comment by tacaswell** [comment]
oh, we should have talked about this on the call!

**Comment by jklymak** [comment]
I'd backport it?

**Comment by brunobeltran** [comment]
+1 for backporting

**Comment by tacaswell** [comment]
@meeseeksdev backport to v3.4.x

This will need a bit of extra work from @QuLogic to merge the api change note in.

**Comment by tacaswell** [comment]
sigh, doing the manual backport.

**Pull request merged**